libcudacxx/include/cuda/std/cstring

```diff
@@ -31,10 +32,22 @@
 
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
-using ::memcpy;
-using ::memset;
 using ::size_t;
 
+_CCCL_API inline void* memcpy(void* __dest, const void* __src, size_t __count) noexcept
+{
+  _CCCL_ASSERT(::cuda::__is_valid_address_range(__src, __count), "memcpy: source range is invalid");
+  _CCCL_ASSERT(::cuda::__is_valid_address_range(__dest, __count), "memcpy: destination range is invalid");
+  _CCCL_ASSERT(::cuda::__are_ptrs_overlapping(__src, __dest, __count), "memcpy: source and destination overlap");
+  return ::memcpy(__dest, __src, __count);
```

Comment:
kinda naive question for my own understanding:

Our cuda version of `memset`/`memcopy` is basically appending device memory preconditions but the implementation still relies on the host compiler's implementation? 

In other words there is not a special way that we should be copying/setting bytes on device memory as long as address ranges and pointers are valid?